be/src/olap/generic_iterators.cpp

```diff
@@ -0,0 +1,346 @@
+    // Before call this function, Client must assure that
+    // has_remaining() return true
+    RowBlockRow current_row() const {
+        return RowBlockRow(&_block, _index_in_block);
+    }
+
+    // Advance internal row index to next valid row
+    // Return error if error happens
+    // Don't call this when valid() is false, action is undefined
+    Status advance();
+
+    // Return if has remaining data in this context.
+    // Only when this function return true, current_row()
+    // will return a valid row
+    bool valid() const { return _valid; }
+
+private:
+    // Load next block into _block
+    Status _load_next_block();
+
+private:
+    RowwiseIterator* _iter;
+    Arena _arena;
+    // used to store data load from iterator
+    RowBlockV2 _block;
+
+    bool _valid = false;
+    size_t _index_in_block = 0;
+};
+
+Status MergeContext::init(const StorageReadOptions& opts) {
+    RETURN_IF_ERROR(_iter->init(opts));
+    RETURN_IF_ERROR(_load_next_block());
+    return Status::OK();
+}
+
+Status MergeContext::advance() {
+    // NOTE: we increase _index_in_block directly to valid one check
+    _index_in_block++;
+    do {
+        for (; _index_in_block < _block.num_rows(); ++_index_in_block) {
+            // TODO(zc): we can skip rows that is fitered by conjunts here
+            // Now we return directly
+            return Status::OK();
+        }
+        // current batch has no data, load next batch
+        RETURN_IF_ERROR(_load_next_block());
+    } while (_valid);
+    return Status::OK();
+}
+
+Status MergeContext::_load_next_block() {
+    Status st;
+    do {
+        st = _iter->next_batch(&_block);
+        if (!st.ok()) {
+            _valid = false;
+            if (st.is_end_of_file()) {
+                return Status::OK();
+            } else {
+                return st;
+            }
+        }
+    } while (_block.num_rows() == 0);
+    _index_in_block = 0;
+    _valid = true;
+    return Status::OK();
+}
+
+struct MergeContextComaprator {
+    MergeContextComaprator(SchemaV2* schema) : _schema(schema) { }
+    bool operator()(const MergeContext* lhs, const MergeContext* rhs) const {
+        auto lhs_row = lhs->current_row();
+        auto rhs_row = rhs->current_row();
+
+        return _schema->compare(lhs_row, rhs_row) > 0;
+    }
+private:
+    SchemaV2* _schema;
+};
+
+class MergeIterator : public RowwiseIterator {
+public:
+    // Iterators' ownership it transfered to this class.
+    // This class will delete all iterators when destructs
+    // Client should not use iterators any more.
+    MergeIterator(std::vector<RowwiseIterator*> iters)
+        : _origin_iters(std::move(iters)) {
+    }
+
+    ~MergeIterator() override {
+        for (auto iter : _origin_iters) {
+            delete iter;
+        }
+        for (auto ctx : _merge_ctxs) {
+            delete ctx;
+        }
+    }
+    Status init(const StorageReadOptions& opts) override;
+    Status next_batch(RowBlockV2* block) override;
```

Comment:
next_block or next_batch?

Reply:
I think batch is better, Because in our storage there is too many things named with block. So we use batch just to mean many rows.